Objects/call.c

```diff
@@ -181,6 +181,13 @@ _PyObject_MakeTpCall(PyThreadState *tstate, PyObject *callable,
      * temporary dictionary for keyword arguments (if any) */
     ternaryfunc call = Py_TYPE(callable)->tp_call;
     if (call == NULL) {
+        if (Py_TYPE(callable) == &PyModule_Type) {
+            PyObject* mod_call = PyObject_GetAttr(callable, &_Py_ID(__call__));
+            if (mod_call != NULL) {
+                return _PyObject_MakeTpCall(tstate, mod_call, args, nargs, keywords);
+            }
```

Comment:
The general convention is to add something to a type object's `tp_call` or `tp_vectorcall_offset` slot. In this case, you would want to modify `Object/moduleobject.c` to have a simple wrapper function. The wrapper function would do the same as you're doing now (looking up `__call__`).

Reply:
I was originally going down this path, but seems to result in the undesired situation where it marks *all* modules as callable, similar to functions/methods/types, and every module object then has a default `__call__` property. That also has the effect of making `callable(module)` always true, even if `module()` immediately raises `TypeError: 'module' object is not callable`.

Quoting my commit message for 524613cdfbe33c1ad5d1197b61786f1960656f5f:

> This is implemented this way, rather than by adding a `tp_call` method
directly to the base module type, as not all modules will be callable,
and the existence of a method wrapper makes it difficult to determine
if the module in question is actually callable, or just has a method
wrapper. This eliminates the need for isinstance checks and special
error handling in said method wrapper to make non-callable modules
continue to appear as non-callable objects.

Is there a better way to do this that doesn't have this downside?

Reply:
In that case, you can add the tp_call, but special-case module types in `callable`, with the same checks you're doing now.